### DISCUSSION TITLE
Salvo way of dependency injection

### DISCUSSION THREAD

**Discussion by CharlieBytesX** [salvo-rs/salvo#668] [Q&A]
Hey, what's the salvo way of managing dependency injection?

I have managed to do it with global OnceCell and with the Affix
With OnceCell it works but don't know clearly the effects in concurrency.
With affix is bad because the "AppState" struct (Or whatever you want to call the struct containing your services or any dependencies you want to inject) is being copied and dropped on each request

I have read that axum and actix, have "withstate" functions on the router, and understood that the struct is copied only once per thread at the beginning of execution.

What's the better way to manage it?

**Comment by GrxE** [comment]
For multithreaded contexts I use `once_cell::sync::OnceCell`. See: [docs](https://docs.rs/once_cell/1.19.0/once_cell/)

**Comment by chrislearn** [comment]
You can just save data in your handler.

**Reply by sidoqi** [reply]
Can you elaborate more about storing data in handler? Any example?

**Comment by chrislearn** [comment]
Any document said "withstate" the struct is copied only once per thread at the beginning of execution?

**Comment by CharlieBytesX** [comment]
Sorry, I finished to read the actix and axum documentation and don't mention copy once per thread.

Actix:
It uses internally dyn to make it cheap to copy

[Setting Data](https://docs.rs/actix-web/latest/actix_web/web/struct.Data.html#setting-data)
Data is set using the app_data methods on App, Scope, and Resource. If data is wrapped in this Data type for those calls, it can be used as an extractor.

Note that Data should be constructed outside the HttpServer::new closure if shared, potentially mutable state is desired. Data is cheap to clone; internally, it uses an Arc.

See also [App::app_data](https://docs.rs/actix-web/latest/actix_web/struct.App.html#method.app_data), [Scope::app_data](https://docs.rs/actix-web/latest/actix_web/struct.Scope.html#method.app_data), and [Resource::app_data](https://docs.rs/actix-web/latest/actix_web/struct.Resource.html#method.app_data).

Axum:
In axum the same is done, but explicitly

<img width="1552" alt="image" src="https://github.com/salvo-rs/salvo/assets/62361574/601de44f-1477-4aa4-a074-27b6423056f1">



Whats the best approach to do something like this? (not the structure or wrapping function, I mean the best way to inject different services to the handlers)
Golang code:

main:
package main

import (
	"fmt"
	"os"
	"strconv"
	"uc-tests/internal/adapter/handler/http"
	"uc-tests/internal/adapter/repository/gorm"
	nextJSRepository "uc-tests/internal/adapter/repository/nextjs"
	"uc-tests/internal/core/services"

	chi "github.com/go-chi/chi/v5"
	_ "github.com/joho/godotenv/autoload"
)

func main() {
	fmt.Fprintf(os.Stderr, "your own program name is %s \n", os.Args[0])

	r := chi.NewRouter()

	// NEXTJS DEV SERVER
	devServerURL := os.Getenv("UI_DEV_SERVER")
	if devServerURL == "" {
		panic("Missing UI_DEV_SERVER")

	}

	//Files Repo
	filesRepo := nextJSRepository.NewNextJSRepository(devServerURL)
	assetService := services.NewAssetProviderService(filesRepo)

	//Admin Auth
	sqliteAdminUserDbPath := os.Getenv("SQLITE_USER_DB")
	sqliteAdminSessionDbPath := os.Getenv("SQLITE_SESSION_DB")

	sqliteAdminSessionRepo := gorm.NewSqliteSessionRepository(sqliteAdminSessionDbPath)
	sqliteAdminUserRepo := gorm.NewSqliteAdminUserRepository(sqliteAdminUserDbPath)
	adminAuthService := services.NewAdminAuthService(sqliteAdminSessionRepo, sqliteAdminUserRepo)

	//Test Service
	sqliteTestDB := os.Getenv("SQLITE_USER_DB")
	testRepository := gorm.NewTestSqliteRepository(sqliteTestDB)
	testService := services.NewTestManagementService(testRepository)

	port := os.Getenv("SERVER_PORT")
	if port == "" {
		panic("Missing port env")

	}

	dev, err := strconv.ParseBool(os.Getenv("DEV_MODE"))
	if err != nil {
		panic("Missing Dev mode")
	}
	server := http.NewServer(r, dev, adminAuthService, assetService, testService)

	server.RegisterRoutes()
	// Server Actions
	server.RegisterDevUIServer()
	server.Serve(port)
}

Register routes:


func (server Server) registerAdminRoutes() {

	adminAuthController := controllers.NewAdminAuthController(server.adminAuthService, adminProtectedRoutesPrefix, server.assetsProvider)
	testController := controllers.NewAdminTestManagementService(server.testService)

	messageWrapper := NewMessageWrapper()

	//Public Admin Routes
	server.router.Post("/api/public/admin/login", adminAuthController.Login)
	server.router.Get("/public/admin/login", adminAuthController.LoginPageOrRedirect)
	server.router.Get("/message", messageWrapper.readMessage)
	server.router.Get("/message/changed", messageWrapper.changeMessage)

	//Protected Admin Routes
	server.router.Route(adminProtectedRoutesPrefix, func(r chi.Router) {
		r.Use(adminAuthController.AuthMiddleware)
		r.Get("/", adminAuthController.RedirectToApp)
		r.Get("/app", adminAuthController.ServePage)
		r.Get("/app/new_test", adminAuthController.ServePage)
		r.Post("/api/new_test", testController.NewTest)
	})

}

**Comment by CharlieBytesX** [comment]
Also, thanks for maintaining such a cool project!
I really appreciate all the work you have done
It's amazing!
